test/jdk/sun/security/pkcs11/KeyStore/ImportKeyToP12.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2022, Red Hat, Inc.
+ *
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.MessageDigest;
+import java.security.Provider;
+import java.security.Security;
+
+import javax.crypto.spec.SecretKeySpec;
+
+/*
+ * @test
+ * @bug 9999999
+ * @summary test SunPKCS11's password based privacy and integrity
+ *          applied to PKCS#12 keystores
+ * @requires (jdk.version.major >= 8)
+ * @library /test/lib ..
+ * @modules java.base/sun.security.util
+ * @run main/othervm/timeout=30 -Dcom.redhat.fips=false -DNO_DEFAULT=true ImportKeyToP12
```

Comment:
I'm getting this on a FIPS enabled system. Expected?
```
$ java -cp work/classes/ImportKeyToP12.java.d -Djava.security.debug=properties ImportKeyToP12
properties: reading security properties file: /home/sgehwolf/mandrel-fips/mandrel-build/conf/security/java.security
properties: {jdk.jar.disabledAlgorithms=MD2, MD5, RSA keySize < 1024, DSA keySize < 1024, fips.provider.3=SunEC, fips.provider.4=SunJSSE, fips.provider.1=SunPKCS11 ${java.home}/conf/security/nss.fips.cfg, fips.provider.2=SUN, jdk.security.legacyAlgorithms=SHA1, RSA keySize < 2048, DSA keySize < 2048, crypto.policy=unlimited, jceks.key.serialFilter=java.base/java.lang.Enum;java.base/java.security.KeyRep;java.base/java.security.KeyRep$Type;java.base/javax.crypto.spec.SecretKeySpec;!*, login.configuration.provider=sun.security.provider.ConfigFile, security.overridePropertiesFile=true, jdk.tls.legacyAlgorithms=NULL, anon, RC4, DES, 3DES_EDE_CBC, security.provider.7=SunSASL, security.provider.8=XMLDSig, security.provider.9=SunPCSC, jdk.security.caDistrustPolicies=SYMANTEC_TLS, security.provider.1=SUN, security.provider.2=SunRsaSign, security.provider.3=SunEC, security.provider.4=SunJSSE, networkaddress.cache.negative.ttl=10, jdk.tls.alpnCharset=ISO_8859_1, security.provider.5=SunJCE, security.provider.6=SunJGSS, ssl.KeyManagerFactory.algorithm=SunX509, ssl.TrustManagerFactory.algorithm=PKIX, fips.provider.5=SunJCE, fips.provider.6=SunRsaSign, policy.allowSystemProperty=true, jdk.io.permissionsUseCanonicalPath=false, package.access=sun.misc.,sun.reflect., package.definition=sun.misc.,sun.reflect., security.provider.12=SunPKCS11, policy.provider=sun.security.provider.PolicyFile, policy.url.1=file:${java.home}/conf/security/java.policy, policy.url.2=file:${user.home}/.java.policy, securerandom.source=file:/dev/random, jdk.certpath.disabledAlgorithms=MD2, MD5, SHA1 jdkCA & usage TLSServer, RSA keySize < 1024, DSA keySize < 1024, EC keySize < 224, jdk.tls.disabledAlgorithms=SSLv3, TLSv1, TLSv1.1, RC4, DES, MD5withRSA, DH keySize < 1024, EC keySize < 224, 3DES_EDE_CBC, anon, NULL, policy.ignoreIdentityScope=false, keystore.type.compat=true, security.provider.11=JdkSASL, security.provider.10=JdkLDAP, jdk.sasl.disabledMechanisms=, fips.keystore.type=PKCS11, sun.security.krb5.maxReferrals=5, security.useSystemPropertiesFile=true, jdk.tls.keyLimits=AES/GCM/NoPadding KeyUpdate 2^37, jdk.xml.dsig.secureValidationPolicy=disallowAlg http://www.w3.org/TR/1999/REC-xslt-19991116,disallowAlg http://www.w3.org/2001/04/xmldsig-more#rsa-md5,disallowAlg http://www.w3.org/2001/04/xmldsig-more#hmac-md5,disallowAlg http://www.w3.org/2001/04/xmldsig-more#md5,disallowAlg http://www.w3.org/2000/09/xmldsig#sha1,disallowAlg http://www.w3.org/2000/09/xmldsig#dsa-sha1,disallowAlg http://www.w3.org/2000/09/xmldsig#rsa-sha1,disallowAlg http://www.w3.org/2007/05/xmldsig-more#sha1-rsa-MGF1,disallowAlg http://www.w3.org/2001/04/xmldsig-more#ecdsa-sha1,maxTransforms 5,maxReferences 30,disallowReferenceUriSchemes file http https,minKeySize RSA 1024,minKeySize DSA 1024,minKeySize EC 224,noDuplicateIds,noRetrievalMethodLoops, securerandom.drbg.config=, sun.security.krb5.disableReferrals=false, keystore.type=pkcs12, securerandom.strongAlgorithms=NativePRNGBlocking:SUN,DRBG:SUN, policy.expandProperties=true, krb5.kdc.bad.policy=tryLast}
properties: reading system security properties file /etc/crypto-policies/back-ends/java.config
properties: {jdk.jar.disabledAlgorithms=MD2, MD5, RSA keySize < 1024, DSA keySize < 1024, fips.provider.3=SunEC, fips.provider.4=SunJSSE, fips.provider.1=SunPKCS11 ${java.home}/conf/security/nss.fips.cfg, fips.provider.2=SUN, jdk.security.legacyAlgorithms=SHA1, RSA keySize < 2048, DSA keySize < 2048, crypto.policy=unlimited, jceks.key.serialFilter=java.base/java.lang.Enum;java.base/java.security.KeyRep;java.base/java.security.KeyRep$Type;java.base/javax.crypto.spec.SecretKeySpec;!*, login.configuration.provider=sun.security.provider.ConfigFile, security.overridePropertiesFile=true, jdk.tls.legacyAlgorithms=, security.provider.7=SunSASL, security.provider.8=XMLDSig, security.provider.9=SunPCSC, jdk.security.caDistrustPolicies=SYMANTEC_TLS, security.provider.1=SUN, security.provider.2=SunRsaSign, security.provider.3=SunEC, security.provider.4=SunJSSE, networkaddress.cache.negative.ttl=10, jdk.tls.alpnCharset=ISO_8859_1, security.provider.5=SunJCE, security.provider.6=SunJGSS, ssl.KeyManagerFactory.algorithm=SunX509, ssl.TrustManagerFactory.algorithm=PKIX, fips.provider.5=SunJCE, fips.provider.6=SunRsaSign, policy.allowSystemProperty=true, jdk.io.permissionsUseCanonicalPath=false, package.access=sun.misc.,sun.reflect., package.definition=sun.misc.,sun.reflect., security.provider.12=SunPKCS11, policy.provider=sun.security.provider.PolicyFile, policy.url.1=file:${java.home}/conf/security/java.policy, policy.url.2=file:${user.home}/.java.policy, securerandom.source=file:/dev/random, jdk.certpath.disabledAlgorithms=MD2, SHA1, MD5, DSA, RSA keySize < 2048, jdk.tls.disabledAlgorithms=DH keySize < 2048, TLSv1.1, TLSv1, SSLv3, SSLv2, RSAPSK, TLS_RSA_WITH_AES_256_CBC_SHA256, TLS_RSA_WITH_AES_256_CBC_SHA, TLS_RSA_WITH_AES_128_CBC_SHA256, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_256_GCM_SHA384, TLS_RSA_WITH_AES_128_GCM_SHA256, DHE_DSS, RSA_EXPORT, DHE_DSS_EXPORT, DHE_RSA_EXPORT, DH_DSS_EXPORT, DH_RSA_EXPORT, DH_anon, ECDH_anon, DH_RSA, DH_DSS, ECDH, 3DES_EDE_CBC, DES_CBC, RC4_40, RC4_128, DES40_CBC, RC2, HmacMD5, policy.ignoreIdentityScope=false, keystore.type.compat=true, security.provider.11=JdkSASL, security.provider.10=JdkLDAP, jdk.sasl.disabledMechanisms=, fips.keystore.type=PKCS11, sun.security.krb5.maxReferrals=5, security.useSystemPropertiesFile=true, jdk.tls.keyLimits=AES/GCM/NoPadding KeyUpdate 2^37, jdk.xml.dsig.secureValidationPolicy=disallowAlg http://www.w3.org/TR/1999/REC-xslt-19991116,disallowAlg http://www.w3.org/2001/04/xmldsig-more#rsa-md5,disallowAlg http://www.w3.org/2001/04/xmldsig-more#hmac-md5,disallowAlg http://www.w3.org/2001/04/xmldsig-more#md5,disallowAlg http://www.w3.org/2000/09/xmldsig#sha1,disallowAlg http://www.w3.org/2000/09/xmldsig#dsa-sha1,disallowAlg http://www.w3.org/2000/09/xmldsig#rsa-sha1,disallowAlg http://www.w3.org/2007/05/xmldsig-more#sha1-rsa-MGF1,disallowAlg http://www.w3.org/2001/04/xmldsig-more#ecdsa-sha1,maxTransforms 5,maxReferences 30,disallowReferenceUriSchemes file http https,minKeySize RSA 1024,minKeySize DSA 1024,minKeySize EC 224,noDuplicateIds,noRetrievalMethodLoops, securerandom.drbg.config=, sun.security.krb5.disableReferrals=false, keystore.type=pkcs12, securerandom.strongAlgorithms=NativePRNGBlocking:SUN,DRBG:SUN, policy.expandProperties=true, krb5.kdc.bad.policy=tryLast}
properties: Calling getSystemFIPSEnabled (libsystemconf)...
properties: getSystemFIPSEnabled: calling SECMOD_GetSystemFIPSEnabled
properties: getSystemFIPSEnabled: SECMOD_GetSystemFIPSEnabled returned 0x1
properties: Call to getSystemFIPSEnabled (libsystemconf) returned: true
properties: FIPS mode detected
properties: Removing provider: security.provider.7=SunSASL
properties: Removing provider: security.provider.8=XMLDSig
properties: Removing provider: security.provider.9=SunPCSC
properties: Removing provider: security.provider.1=SUN
properties: Removing provider: security.provider.2=SunRsaSign
properties: Removing provider: security.provider.3=SunEC
properties: Removing provider: security.provider.4=SunJSSE
properties: Removing provider: security.provider.5=SunJCE
properties: Removing provider: security.provider.6=SunJGSS
properties: Removing provider: security.provider.12=SunPKCS11
properties: Removing provider: security.provider.11=JdkSASL
properties: Removing provider: security.provider.10=JdkLDAP
properties: Adding provider 1: security.provider.1=SunPKCS11 ${java.home}/conf/security/nss.fips.cfg
properties: Adding provider 2: security.provider.2=SUN
properties: Adding provider 3: security.provider.3=SunEC
properties: Adding provider 4: security.provider.4=SunJSSE
properties: Adding provider 5: security.provider.5=SunJCE
properties: Adding provider 6: security.provider.6=SunRsaSign
properties: FIPS mode default keystore.type = PKCS11
properties: FIPS mode javax.net.ssl.keyStore = NONE
properties: FIPS mode javax.net.ssl.trustStoreType = pkcs12
properties: FIPS support enabled with plain key support
properties: FIPS support enabled.
SunPKCS11: SunPKCS11-NSS-FIPS
=========================================================================
Cipher PBE: PBEWithHmacSHA1AndAES_128
Mac PBE: HmacPBESHA1
Exception in thread "main" java.io.IOException: Integrity check failed: java.security.NoSuchAlgorithmException: Algorithm HmacPBESHA1 not available
	at java.base/sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2224)
	at java.base/sun.security.util.KeyStoreDelegator.engineLoad(KeyStoreDelegator.java:221)
	at java.base/java.security.KeyStore.load(KeyStore.java:1473)
	at ImportKeyToP122.testWith(ImportKeyToP12.java:124)
	at ImportKeyToP122.main(ImportKeyToP12.java:86)
	at ImportKeyToP122.main(ImportKeyToP12.java:137)
	at ImportKeyToP12.main(ImportKeyToP12.java:49)
Caused by: java.security.NoSuchAlgorithmException: Algorithm HmacPBESHA1 not available
	at java.base/javax.crypto.Mac.getInstance(Mac.java:190)
	at java.base/sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2198)
	... 6 more
```

Reply:
Looks like a test problem. On a FIPS system `HmacPBESHA1` is only provided by `SunPKCS11-NSS-FIPS` but that provider is removed prior the integrity check here:
https://github.com/franferrax/jdk/blob/RH2048582/test/jdk/sun/security/pkcs11/KeyStore/ImportKeyToP12.java#L124

Reply:
@jerboaa: I think it's worth giving a bit of context here.

For all the tests except this one, we've taken the following approach regarding assertion data: we try to obtain the
expected data (derived key, hash-based message authentication code, or cipher text) from _SunJCE_ (our "trusted"
implementation), if _SunJCE_ can't be instantiated, we fall back to test-embedded/hardcoded assertion data.

The idea behind this is to self-document the way we obtain the assertion data, allowing to easily create new tests [\[&dagger;\]](#d994eb6f8ced519026769aa02596c19e)
that also work in FIPS-mode (where _SunJCE_ is not available). This way, we can have upstream-ready tests (they are
just _SunPKCS11_ tests, to be proposed along with most of this PR work [\[&ddagger;\]](#a831e98775a761abc013ddbeb3214e48)), that also run downstream and in FIPS mode.

In particular, for `ImportKeyToP12.java`, we faced the following issues (@martinuy knows the full details, since he developed this one):
* We didn't want to hardcode a certificate or a whole keystore
* Generating a certificate on-the-fly would require a significant amount of code (generating its key was the easy part)
* Implementing the test as a `SecmodTest` (using an existing certificate from `test/jdk/sun/security/pkcs11/Secmod`) wouldn't work upstream because the certificate's key have the `CKA_SENSITIVE` attribute (and upstream, we don't have the work of #1, nor it is upstreamable)
* PKCS#12 can't store a certificate's key if there is no certificate associated to it

Finally, we decided to just store a `SecretKey` (symmetric, so no certificates involved) and to rely on available providers (typically _SunJCE_) to make a cross-check: create a keystore with _SunPKCS11_'s crypto, then read it with other available providers' crypto. That's why _SunPKCS11_ is removed from the providers, and why this test isn't meant to be run in FIPS mode. Please note that the test still exercises the _SunPKCS11_ PBA and PBE primitives used for PKCS#12 keystores, which are the ones used when in FIPS mode.

#### <a id="d994eb6f8ced519026769aa02596c19e"></a>\[&dagger;\] New test creation steps
1. Develop the test with an empty test-embedded fallback assertion data
1. Run it in a FIPS-disabled OpenJDK (so _SunJCE_ is available), and make sure it passes
1. Copy the printed _SunPKCS11_'s actual data (but asserted to be equal to _SunJCE_'s expected data)
1. Paste this data as the test-embedded fallback assertion data
1. Now we are able to run the test in both FIPS and non-FIPS

#### <a id="a831e98775a761abc013ddbeb3214e48"></a>\[&ddagger;\] So far, non-upstreamable bits include
* [Little changes in `SunJCE.java`](/rh-openjdk/jdk/pull/2/commits/8dc75ca65657b075f1de799d29066f67b5b5ad28#diff-f4b30a0e39eaadf9bb4df92dcf04ff8c80f8f54f0e2b79fff5def9fb7eef2eb5)
* NSS FIPS-tests workaround by invoking `java.security.Security.getProviders()`, we must propose `TestClass2` as upstream's `TestClass`, for `TestClass` in: ([`PBECipher`](/rh-openjdk/jdk/pull/2/commits/8dc75ca65657b075f1de799d29066f67b5b5ad28#diff-f7a7547c02cc465a5f7992810c629ffa5da5c5ceeddb40e11dfc173bcdaf84a1R52), [`ImportKeyToP12`](/rh-openjdk/jdk/pull/2/commits/8dc75ca65657b075f1de799d29066f67b5b5ad28#diff-d39e708032485b874b14df66f73ec241f844fc73dff152c0c849d33f964a0612R49), [`PBAMac`](/rh-openjdk/jdk/pull/2/commits/8dc75ca65657b075f1de799d29066f67b5b5ad28#diff-285e8731b40dea0132e259250894fb72c9d5121976474c3c1be21b7b5ac4cc18R49), [`TestPBKD`](/rh-openjdk/jdk/pull/2/commits/8dc75ca65657b075f1de799d29066f67b5b5ad28#diff-350f80ceb87cc00b47cfae84d9d909ab66b1a62f0988df6e39d7f9b194c26440R49))

Reply:
OK, I understand. I've worked around this (on a FIPS system) by doing the verification steps in a separate Java program that is running in non-fips mode which seems a nice inter-op test too.

Reply:
> #### \[&ddagger;\] So far, non-upstreamable bits include
> * [Little changes in `SunJCE.java`](/rh-openjdk/jdk/commit/8dc75ca65657b075f1de799d29066f67b5b5ad28#diff-f4b30a0e39eaadf9bb4df92dcf04ff8c80f8f54f0e2b79fff5def9fb7eef2eb5)
> * NSS FIPS-tests workaround by invoking `java.security.Security.getProviders()`, we must propose `TestClass2` as upstream's `TestClass`, for `TestClass` in: ([`PBECipher`](/rh-openjdk/jdk/commit/8dc75ca65657b075f1de799d29066f67b5b5ad28#diff-f7a7547c02cc465a5f7992810c629ffa5da5c5ceeddb40e11dfc173bcdaf84a1R52), [`ImportKeyToP12`](/rh-openjdk/jdk/commit/8dc75ca65657b075f1de799d29066f67b5b5ad28#diff-d39e708032485b874b14df66f73ec241f844fc73dff152c0c849d33f964a0612R49), [`PBAMac`](/rh-openjdk/jdk/commit/8dc75ca65657b075f1de799d29066f67b5b5ad28#diff-285e8731b40dea0132e259250894fb72c9d5121976474c3c1be21b7b5ac4cc18R49), [`TestPBKD`](/rh-openjdk/jdk/commit/8dc75ca65657b075f1de799d29066f67b5b5ad28#diff-350f80ceb87cc00b47cfae84d9d909ab66b1a62f0988df6e39d7f9b194c26440R49))

`SunJCE.java` is no longer touched, after rebasing to include #16 changes.